pkg/proxy/proxy.go

```diff
@@ -136,130 +145,191 @@ func (p *Proxy) Run(stopCh <-chan struct{}) (<-chan struct{}, error) {
+				allAuthFound = true
+				break
+			}
+		}
+		if !allAuthFound {
+			groups = append(groups, authuser.AllAuthenticated)
+		}
 
-	// If client IP user extra header option set then append the remote client
-	// address.
-	if p.options.ExtraUserHeadersClientIPEnabled {
-		klog.V(6).Infof("adding impersonate extra user header %s: %s (%s)",
-			UserHeaderClientIPKey, req.RemoteAddr, reqCpy.RemoteAddr)
+		extra := user.GetExtra()
 
-		extra[UserHeaderClientIPKey] = append(extra[UserHeaderClientIPKey], req.RemoteAddr)
-	}
+		if extra == nil {
+			extra = make(map[string][]string)
+		}
 
-	// Add custom extra user headers to impersonation request
-	for k, vs := range p.options.ExtraUserHeaders {
-		for _, v := range vs {
+		// If client IP user extra header option set then append the remote client
+		// address.
+		if p.config.ExtraUserHeadersClientIPEnabled {
 			klog.V(6).Infof("adding impersonate extra user header %s: %s (%s)",
-				k, v, reqCpy.RemoteAddr)
+				UserHeaderClientIPKey, req.RemoteAddr, req.RemoteAddr)
 
-			extra[k] = append(extra[k], v)
+			extra[UserHeaderClientIPKey] = append(extra[UserHeaderClientIPKey], req.RemoteAddr)
 		}
-	}
 
-	// Set impersonation header using authenticated user identity.
-	conf := transport.ImpersonationConfig{
-		UserName: user.GetName(),
-		Groups:   groups,
-		Extra:    extra,
-	}
+		// Add custom extra user headers to impersonation request.
+		for k, vs := range p.config.ExtraUserHeaders {
+			for _, v := range vs {
+				klog.V(6).Infof("adding impersonate extra user header %s: %s (%s)",
+					k, v, req.RemoteAddr)
+
+				extra[k] = append(extra[k], v)
+			}
+		}
 
-	rt := transport.NewImpersonatingRoundTripper(conf, p.clientTransport)
+		conf := &transport.ImpersonationConfig{
+			UserName: user.GetName(),
+			Groups:   groups,
+			Extra:    extra,
+		}
 
-	// push request through round trippers to the API server
-	return rt.RoundTrip(reqCpy)
+		// Add the impersonation configuration to the context.
+		req = req.WithContext(context.WithImpersonationConfig(req.Context(), conf))
+		handler.ServeHTTP(rw, req)
+	})
 }
 
-func (p *Proxy) tokenReview(req *http.Request) (*http.Response, error) {
+func (p *Proxy) reviewToken(rw http.ResponseWriter, req *http.Request) bool {
 	klog.V(4).Infof("attempting to validate a token in request using TokenReview endpoint(%s)",
 		req.RemoteAddr)
 
 	ok, err := p.tokenReviewer.Review(req)
-	// no error so passthrough the request
+
+	// No error and ok so passthrough the request
 	if err == nil && ok {
 		klog.V(4).Infof("passing request with valid token through (%s)",
 			req.RemoteAddr)
-		// Don't set impersonation headers and pass through without proxy auth
-		// and headers still set
-		return p.noAuthClientTransport.RoundTrip(req)
+
+		return true
 	}
 
 	if err != nil {
 		klog.Errorf("unable to authenticate the request via TokenReview due to an error (%s): %s",
 			req.RemoteAddr, err)
 	}
 
-	return nil, errUnauthorized
+	return false
```

Comment:
This would be easier to read if the 'happy path' ended here, ie:

```
ok, err := p.tokenReviewer.Review(req)
if err != nil {
  klog.Errorf("unable to authenticate the request via TokenReview due to an error (%s): %s", req.RemoteAddr, err)
  return false
}
if !ok {
  klog.V(4).Infof("user not authorized to make request (%s)", req.RemoteAddr)
  return false
}

klog.Errorf("unable to authenticate the request via TokenReview due to an error (%s): %s", req.RemoteAddr, err)
return true
```

Reply:
Yep, I agree